### PR TITLE
fix(client): sign up for email list directly

### LINKED ISSUE #40772: Ability to sign up for email list directly
Currently if you visit https://www.freecodecamp.org/email-sign-up/ it looks as though you can sign up for the email list. But signing up here will trigger an error message.

Here is how I propose we remedy this:

If you attempt to sign up for the mailing list and you are not signed in, you are asked to sign in.

But since https://www.freecodecamp.org/email-sign-up/ is the URL we normally send new people to when they are first signing up, we don't want to send these people back to this URL, as this would be confusing and most people would not realize they need to scroll down and click "yes" a second time.

Instead we want to send them to /learn and show a confirmation message up top that they are now signed up for the weekly email.

----

Edit by @raisedadead

Here is what we want to do with this from https://github.com/freeCodeCamp/freeCodeCamp/pull/45508#issuecomment-1077405988:

So, I looked into this, and the alternative way seems to me like we need a dedicated landing page for users we send to the "sign-up for the mailing list" feature. 

There are a few things to note:

- The path `/email-sign-up` has to be used for the feature. This path is already in the wild (looks logical, btw).
- The landing page needs to have a CTA and verbiage along the lines of "benefits of signing up for the weekly links". The text can be verbatim from the current page & add a bit more, IDK.
- The mailing list landing page can implicitly set the mailing list consent flag. We will need to add some query parameters when we redirect through the sign-up flow, which will show users a thank-you page when the flow is complete.
- We have been abusing the `/email-sign-up` for something else. We use this route to comply with GDPR; in other words, it's an interstitial (just like our donation modals) for setting the `acceptPrivacyTerms` requirement. We instead need to create a brand new page say: `/welcome-on-board` for this and route users to it when they sign up as regular learners. This page shows the current content from the `/email-sign-up` about pacing oneself.

**TL;DR:** Both pages do one common thing, i.e., implicitly set the `acceptPrivacyTerms` and show their target audience mutually exclusive information (pacing oneself vs. mailing list).

### PR DESCRIPTION
**Pull request by raisedadead**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ref #40772

image

### CODE REVIEW

File: client/src/pages/email-sign-up.tsx [hunk 1]
```diff
@@ -59,28 +69,11 @@ function AcceptPrivacyTerms({
...
-  return acceptedPrivacyTerms ? (
-    <RedirectToLearn />
-  ) : (
-    <>
-      <Helmet>
-        <title>{t('misc.email-signup')} | freeCodeCamp.org</title>
-      </Helmet>
-      <Grid>
+  function renderEmailListOptin(isSignedIn: boolean) {
+    const href = isSignedIn ? `${homeLocation}/learn` : `${apiLocation}/signin`;
```
Comment (raisedadead) [inline client/src/pages/email-sign-up.tsx:73]:
```suggestion

```

File: client/src/pages/email-sign-up.tsx [hunk 2]
```diff
@@ -105,6 +98,50 @@ function AcceptPrivacyTerms({
...
+    } else {
+      return (
+        <Col md={8} mdOffset={2} sm={10} smOffset={1} xs={12}>
+          <ButtonSpacer />
+          <Button
+            block={true}
+            bsSize='lg'
+            bsStyle='primary'
+            className='big-cta-btn'
+            href={href}
```
Comment (raisedadead) [inline client/src/pages/email-sign-up.tsx:112]:
```suggestion
            href=`${apiLocation}/signin`
```

File: client/src/pages/email-sign-up.tsx [hunk 3]
```diff
@@ -9,21 +9,29 @@ import { createSelector } from 'reselect';
 import IntroDescription from '../components/Intro/components/IntroDescription';
 import createRedirect from '../components/create-redirect';
 import { ButtonSpacer, Spacer } from '../components/helpers';
+import envData from '../../../config/env.json';
 
-import { acceptTerms, userSelector } from '../redux';
+import { acceptTerms, userSelector, isSignedInSelector } from '../redux';
 
-import './email-sign-up.css';
+const { apiLocation, homeLocation } = envData;
```
Comment (raisedadead) [inline client/src/pages/email-sign-up.tsx:16]:
```suggestion

```

File: client/src/pages/email-sign-up.tsx [hunk 4]
```diff
@@ -9,21 +9,29 @@ import { createSelector } from 'reselect';
 import IntroDescription from '../components/Intro/components/IntroDescription';
 import createRedirect from '../components/create-redirect';
 import { ButtonSpacer, Spacer } from '../components/helpers';
+import envData from '../../../config/env.json';
```
Comment (raisedadead) [inline client/src/pages/email-sign-up.tsx:12]:
```suggestion
import { apiLocation } from '../../../config/env.json';
```

**Review by raisedadead — COMMENTED** [review]
Remove unneeded code copied from elsewhere.

**Comment by raisedadead** [comment]
I would appreciate some ideas on the original use story from the original issue #40772, i.e.:

> ...we don't want to send these people back to this URL,...

This PR should prevent the abrupt access token error, but it still needs to send them to the explicit opt-in page, which happens to be this very page, resulting in the double visit.

I am inclined to accept this as is.

**Comment by raisedadead** [comment]
@freeCodeCamp/dev-team 

So, I looked into this, and the alternative way seems to me like we need a dedicated landing page for users we send to the "sign-up for the mailing list" feature. 

There are a few things to note:

- The path `/email-sign-up` has to be used for the feature. This path is already in the wild (looks logical, btw).
- The landing page needs to have a CTA and verbiage along the lines of "benefits of signing up for the weekly links". The text can be verbatim from the current page & add a bit more, IDK.
- The mailing list landing page can implicitly set the mailing list consent flag.
We have been abusing the `/email-sign-up` for something else. We use this route to comply with GDPR; in other words, it's an interstitial (just like our donation modals) for setting the `acceptPrivacyTerms` requirement. 
- We need to create a brand new page say: `/welcome-on-board` for this and route users to it when they sign up as regular learners. This page shows the current content from the `/email-sign-up` about pacing oneself.


**TL;DR:** Both pages do one common thing, i.e., implicitly set the `acceptPrivacyTerms` and show their target audience mutually exclusive information (pacing oneself vs. mailing list).

**Parting thoughts:**

If we go ahead with this approach, we can't implement this now and will have to wait for the containerized deployments to go out first. It will be a massive breaking change impacting every user on `/learn` - needing some downtime.

So, I propose we go ahead with this PR as is, that at least fixed the "authorization token needed" bug.

**Review by naomi-lgbt — APPROVED** [review]
Looks good to me.
